lib/Sema/TypeCheckStmt.cpp

```diff
@@ -442,6 +442,32 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
       return RS;
     }
 
+    // If the body consisted of a single return without a result
+    // 
+    //   func foo() -> Int {
+    //     return
+    //   }
+    // 
+    // in parseAbstractFunctionBody the return is given an empty, implicit tuple
+    // as its result
+    //
+    //   func foo() -> Int {
+    //     return ()
+    //   }
+    //
+    // Look for that case and diagnose it as missing return expression.
+    if (!ResultTy->isVoid() && TheFunc->hasSingleExpressionBody()) {
+      auto expr = TheFunc->getSingleExpressionBody();
+      if (expr->isImplicit()) {
+        if (auto tuple = dyn_cast<TupleExpr>(expr)) {
+          if (tuple->getNumElements() == 0 && !ResultTy->isVoid()) {
+            TC.diagnose(RS->getReturnLoc(), diag::return_expr_missing);
+            return nullptr;
```

Comment:
I think you can still return `RS` here, like the other case does. Or does that cause more failures?

Reply:
I thought it result in the earlier location-less failure staying in, but I was confused.  Fixed in https://github.com/apple/swift/pull/25802/commits/3761b02d74721cf0ccedae16ba17c2500ef5637a .